crates/pathfinder/src/state/sync.rs

```diff
@@ -138,24 +121,35 @@ where
 
     loop {
         tokio::select! {
+            l1_event = rx_l1.recv() => match l1_event {
+                Some(update) => {
+                    let tx = db_conn.transaction().context("db tx")?;
+
+                    let l1_l2_head = RefsTable::get_l1_l2_head(&tx)?.map(|x| x.0).unwrap_or_default();
+                    if update.block_number <= l1_l2_head {
+                        tracing::info!(block=update.block_number, "L1 reorg detected");
+                    }
+                    RefsTable::set_l1_l2_head(&tx, Some(StarknetBlockNumber(update.block_number)))?;
+                    tracing::info!(block=update.block_number, "L1-L2 head update");
+
+                    if let Some(block) = StarknetBlocksTable::get(&tx, StarknetBlocksBlockId::Number(StarknetBlockNumber(update.block_number)))? {
+                        if block.root.0.as_ref() != update.global_root.as_bytes() {
+                            tracing::warn!(block=update.block_number, "Block root mismatch (L1={}, DB={})", update.global_root, block.root);
+                            // TODO(SM): shut down? ignore?
```

Comment:
Definitely not shut down 😬 Ideally we would want to search for a block which does still match L2 and update it to that. 

However, this would require a restructuring of how this currently works. This is because we no longer want an auto-L1 poll in this case, but rather want a separate task that finds this matching block. One solution is to change from a separate L1 task thread that polls + events, to just using normal futures, which will change from a normal poll-latest-for-changes to a find-matching-block-in-the-past type.

As mentioned before, probably best for a next PR -- this case is unlikely to occur, so just leaving it at `ignore` for now suffices.

Reply:
ack